---
title: "C-4XP507 (P-0033223)"
subtitle: "Salivary Duct Carcinoma"
output:
  html_document:
    df_print: paged
---
***

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, fig.width=10, fig.height=6)
library(data.table)
library(tidyr)
library(dplyr)
library(reshape2)
library(ggplot2)
library(plotly)
library(scales)
library(kableExtra)
library(DT)
theme_set(theme_bw())
```

```{r read-input, include=FALSE}
sample="C-4XP507"
access_data_analysis_results_dir<-"/juno/work/bergerm1/bergerlab/access_projects/Project_06302_TDM1/access_data_analysis_results_v3"
impact_sample_id<-"P-0033223-T02-IM6"
impact_ccf_file<-"/juno/work/ccs/shared/resources/impact/facets/all/P-00332/P-0033223-T02-IM6_P-0033223-N01-IM6/default/P-0033223-T02-IM6_P-0033223-N01-IM6.ccf.maf"
metadata_file<-"/juno/work/bergerm1/bergerlab/access_projects/Project_06302_TDM1/metadata/for_access_data_analysis.2020-08-26.csv"
treatment_info_file<-"/juno/work/bergerm1/bergerlab/access_projects/Project_06302_TDM1/metadata/tdm1_treatment_info.tsv"
## Possible timepoints in the treatment info file, and their associated colors
timepoints<-c("Pre-Treatment", "On-Treatment", "Disease Progression")
timepointcolors<-c("Khaki","Aquamarine","Light Salmon")
names(timepointcolors)<-timepoints
```

```{r global_functions, include=FALSE}
# Function to get variant colors #
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

# Function to generate VAF plots
vaf_plot <- function(variant_table, xlimits, xbreaks, varcolors, yaccuracy=0.01, log=FALSE, cnadjusted=FALSE){
  tmp<-variant_table
  tmp$VarName<-factor(tmp$VarName,levels=names(varcolors))
  if(log == FALSE){
    if(cnadjusted==FALSE){
      fig<-ggplot(tmp, aes(x=collection_date, y=vaf, col=VarName, hotspot=Hotspot, sampleid=cmo_sample_id_plasma)) +
        scale_y_continuous(labels = scales::percent_format(accuracy = yaccuracy)) +
        ylab("VAF")
      } else{
        fig<-ggplot(tmp, aes(x=collection_date, y=adjustedvaf, vaf=vaf, col=VarName, hotspot=Hotspot, sampleid=cmo_sample_id_plasma)) +
          scale_y_continuous(labels = scales::percent_format(accuracy = yaccuracy)) +
          ylab("Adjusted VAF")
      }
  } else{
    if(cnadjusted==FALSE){
      tmp$vafforlog = tmp$vaf + 0.0001
      fig<-ggplot(tmp, aes(x=collection_date, y=vafforlog, vaf=vaf, col=VarName, hotspot=Hotspot, sampleid=cmo_sample_id_plasma)) +
        scale_y_log10(labels = scales::percent_format(accuracy = yaccuracy)) +
        ylab("VAF")
    } else{
      tmp$vafforlog = tmp$adjustedvaf + 0.0001
      fig<-ggplot(tmp, aes(x=collection_date, y=vafforlog, adjustedvaf=adjustedvaf, vaf=vaf, col=VarName, hotspot=Hotspot, sampleid=cmo_sample_id_plasma)) +
        scale_y_log10(labels = scales::percent_format(accuracy = yaccuracy)) +
        ylab("VAF")
    }
  }
  fig<-fig + geom_point() + 
    geom_line(aes(group=VarName)) + 
    facet_grid(clonality~.) + 
    theme(legend.title=element_blank(), strip.background = element_blank(), axis.text.x = element_text(angle=90)) +
    scale_color_manual(values = varcolors) + 
    scale_x_date(breaks=xbreaks, limits=xlimits) + 
    xlab("")
  if(cnadjusted==FALSE){
    fig<-ggplotly(fig, tooltip=c("collection_date","vaf","VarName","Hotspot","cmo_sample_id_plasma"))
  } else{
    fig<-ggplotly(fig, tooltip=c("collection_date","vaf","adjustedvaf","VarName","Hotspot","cmo_sample_id_plasma"))
  }
  fig<-ggplotly(fig)
  return(fig)
}

# Function to generate treatment plot
treatment_plot <- function(treatment_table, xbreaks, colors, ids=c("collection_date","cmo_sample_id_plasma")){
  # Any column other than the ids columns will be plotted
  tmp<-melt(treatment_table, id=ids)
  tmp$value<-factor(tmp$value, levels=names(colors))
  fig<-ggplot(tmp, aes(x=collection_date, y=variable, fill=value)) +
    geom_tile() + 
    theme(panel.grid=element_blank(), legend.title = element_blank()) + 
    scale_x_date(breaks=xbreaks) +
    scale_fill_manual(values=colors)
  fig<-ggplotly(fig)
  fig<-style(fig, showlegend=FALSE)
  return(fig)
}

cna_plot<-function(cna, xlimits, xbreaks){
  cna$fc<-round(cna$fc,2)
  fig<-ggplot(cna, aes(x=collection_date, sampleid=cmo_sample_id_plasma, gene=Hugo_Symbol)) +
    geom_bar(aes(y=1, fill=fc), stat="identity",width=10) +
    scale_fill_gradient2(low="white",mid="red",high="red3",midpoint=1.5, limits=c(0,max(cna$fc))) +
    theme(legend.title=element_blank(), panel.grid.major.y=element_blank(), strip.background = element_blank(), axis.text.x=element_text(angle=90), axis.ticks.y=element_blank(), axis.text.y=element_blank()) +
    scale_x_date(breaks=xbreaks, limits=xlimits) +
    xlab("") + facet_grid(Hugo_Symbol~.)
  fig<-ggplotly(fig, tooltip=c("collection_date","CNA","cmo_sample_id_plasma","fc","Hugo_Symbol"))
  return(fig)
}

cna_plot_old<-function(cna_table, xlimits, xbreaks){
#  cols<-c("red","blue")
#  names(cols)<-c("AMP","DEL")
  fig<-ggplot(cna, aes(x=collection_date, y=Hugo_Symbol, fill=fc, cna=CNA, sampleid=cmo_sample_id_plasma)) +
    geom_tile(col="white", size=2, width=0.7, height=0.7) +
    scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0) +
    theme(legend.title=element_blank(), strip.background = element_blank(), axis.text.x=element_text(angle=90)) +
    scale_x_date(breaks=xbreaks, limits=xlimits) +
    xlab("") + theme(legend.position="none")
  fig<-ggplotly(fig) # %>% layout(showlegend=F, showcale=F)
  return(fig)
}
print_table<-function(table){
  datatable(table, rownames=FALSE, escape=FALSE, options=list(scrollX=T, autoWidth = TRUE))
}
```

```{r read_input}

## read variant files ##
accesstab<-fread(paste0(access_data_analysis_results_dir, "/results_stringent_combined/",sample,"_table.csv"),na.strings=c("NA",""))

dmp<-fread(impact_ccf_file)[,c("Hugo_Symbol", "Chromosome", "Start_Position", "Reference_Allele", "Tumor_Seq_Allele2", "tcn", "expected_alt_copies", "clonality")]
dmp$Chromosome<-as.character(dmp$Chromosome)


## read metadata files ##
metadata<-fread(metadata_file)[,c("cmo_sample_id_plasma", "sex", "collection_date")]
metadata$collection_date<-as.Date(metadata$collection_date, format="%m/%d/%Y")
treatment<-fread(treatment_info_file)
treatment$collection_date<-as.Date(treatment$collection_date, format="%m/%d/%Y")
treatment<-treatment[grep(sample,treatment$cmo_sample_id_plasma),] %>% distinct()


## ACCESS variants ##
tab1<-accesstab[,c("Hugo_Symbol","Chromosome", "Start_Position", "Variant_Classification", "HGVSp_Short", "Reference_Allele", "Tumor_Seq_Allele2", "Hotspot", "DMP", "CH")]
tab2<-accesstab %>% select(matches("__total$"))
colnames(tab2)<-gsub("___total","",colnames(tab2))
cmo_sample_id_order<-metadata$cmo_sample_id_plasma[order(metadata$collection_date)]
cmo_sample_id_order<-cmo_sample_id_order[cmo_sample_id_order %in% colnames(tab2)]
tab2<-data.frame(tab2,check.names=FALSE)[,cmo_sample_id_order]
tab3<-accesstab %>% select(matches("__DMP_Tumor$"))
colnames(tab3)<-gsub("___DMP_Tumor","",colnames(tab3))

dmpsamples<-colnames(tab3)
tab<-cbind(tab1,tab2)
tab<-cbind(tab,tab3)
tab$VarType<-"SNV_INDEL"
tab[grep("__",tab$Hugo_Symbol),"VarType"]<-"SV"

tab$Hugo_Symbol<- gsub("__","-",tab$Hugo_Symbol)
tab$Chromosome<-as.character(tab$Chromosome)
tabletoprint<-merge(tab, dmp, all.x=TRUE)

## read CNA files ##
all.cna.files <- list.files(path = paste0(access_data_analysis_results_dir,"/CNA_final_call_set"),pattern = sample, full.names=TRUE)
cna<-do.call('rbind', lapply(all.cna.files,fread)) %>% rename(cmo_sample_id_plasma=Tumor_Sample_Barcode)
if( nrow(cna) > 0){
  cna<-merge(cna, metadata,by=c("cmo_sample_id_plasma"))
  cna$collection_date<-as.Date(cna$collection_date, format="%m/%d/%Y")
  cna$fc<-as.numeric(cna$fc)
}
```

```{r prepare_data}

## Get VAFs of variants ##
temp <- melt(tab, id=c("Hugo_Symbol","Chromosome", "Start_Position", "Variant_Classification", "HGVSp_Short", "Reference_Allele", "Tumor_Seq_Allele2", "Hotspot", "DMP", "CH", "VarType",dmpsamples)) %>% separate(col=value, into=c("num1", "num2", "num3"))
temp$vaf = as.numeric(temp$num1)/as.numeric(temp$num2)
temp$total=temp$num2
tempsv<-subset(temp,temp$VarType=="SV")
temp[which(temp$VarType=="SV"),"vaf"] = (as.numeric(tempsv$num1)+as.numeric(tempsv$num2))/as.numeric(tempsv$num3)
temp[which(temp$VarType=="SV"),"total"]=tempsv$num3
temp$Chromosome <- as.character(temp$Chromosome)
temp$vaf=temp$vaf

# merge with DMP
combined<-merge(temp, dmp, all.x=TRUE) %>% rename(cmo_sample_id_plasma = variable)

final<-merge(combined, metadata, by=c("cmo_sample_id_plasma"), all.x=TRUE)
final$collection_date<-as.Date(final$collection_date, format="%m/%d/%Y")
final<-merge(final,treatment,by=c("collection_date","cmo_sample_id_plasma"),all.x=TRUE)

# Create a variant name column
final$VarName=paste(final$Hugo_Symbol, final$HGVSp_Short)

final[is.na(final$HGVSp_Short), "VarName"] <- paste0(final$Hugo_Symbol, " ", final$Chromosome, ":", final$Start_Position, " ", final$Reference_Allele, ">", final$Tumor_Seq_Allele2)[is.na(final$HGVSp_Short)]

final[is.na(final$HGVSp_Short) & nchar(final$Reference_Allele)>5,"VarName"] <- paste0(final$Hugo_Symbol, " ", final$Chromosome, ":", final$Start_Position, " ", substr(final$Reference_Allele,1,3),"..", ">", final$Tumor_Seq_Allele2)[is.na(final$HGVSp_Short)]

final[is.na(final$HGVSp_Short) & nchar(final$Tumor_Seq_Allele2)>5,"VarName"] <- paste0(final$Hugo_Symbol, " ", final$Chromosome, ":", final$Start_Position, " ", final$Reference_Allele,1,3, ">", substr(final$Tumor_Seq_Allele2,1,3),"..")[is.na(final$HGVSp_Short)]

final[which(final$VarType=="SV"),"VarName"]<-final$Hugo_Symbol[which(final$VarType=="SV")]

# Select variant colors

final$vaf<-round(final$vaf,4)
final<-final[!is.na(final$vaf)]
final$total<-as.numeric(final$total)
final<-subset(final, final$total>=100)
vafsum<-aggregate(final$vaf,by=list(VarName=final$VarName),FUN=sum)
final<-subset(final, !final$VarName %in% vafsum[which(vafsum$x==0),"VarName"])
varcolors<-c(gg_color_hue(length(unique(final$VarName))),"black")
names(varcolors)<-c(unique(final$VarName),"average")

tabletoprint<-unique(merge(unique(final[,c("Hugo_Symbol","Chromosome","Start_Position")]), tabletoprint, by=c("Hugo_Symbol","Chromosome","Start_Position")))

final_snv<-subset(final,final$VarType=="SNV_INDEL")
final_snv[is.na(final_snv$clonality),"clonality"]<-"ACCESS-only"

final_snv$clonality<-factor(final_snv$clonality, levels=c("CLONAL","SUBCLONAL","INDETERMINATE","ACCESS-only"))
final_sv<-subset(final,final$VarType=="SV")
final_sv$clonality="ACCESS-only"
final_sv[which(final_sv$DMP=="Signed out"),"clonality"]="IMPACT"

xlimits<-c(min(final$collection_date)-10, max(final$collection_date)+10)
xbreaks<-final$collection_date
```

## Sample information
```{r sample-info-table}
unique(final[,c("cmo_sample_id_plasma","collection_date","Timepoint")]) %>% kable() %>% kable_styling()
```

## SNVs and INDELs{.tabset}
### Linear

```{r plot_snvs_linear}
treatment_table<-unique(final[,c("cmo_sample_id_plasma","collection_date","Timepoint")])
fig1<-treatment_plot(treatment_table, xbreaks, timepointcolors, ids=c("collection_date","cmo_sample_id_plasma"))
fig2<-vaf_plot(final_snv, xlimits, xbreaks, varcolors, yaccuracy=0.01, log=FALSE)
subplot(fig1,fig2,nrows=2,shareX=TRUE, heights=c(0.1,0.9))
```

### Log

```{r plot_snvs_log}
fig2<-vaf_plot(final_snv, xlimits, xbreaks, varcolors, yaccuracy=0.01, log=TRUE)
subplot(fig1,fig2,nrows=2,shareX=TRUE, heights=c(0.1,0.9))
```

### Table
```{r snvs_table}
print_table(subset(tabletoprint,tabletoprint$VarType=="SNV_INDEL"))
```

### Description
* We used FACETS results of `r toString(impact_sample_id)` to identify CLONAL, SUBCLONAL or INDETERMINATE mutations.
* Mutations that were not called in `r toString(impact_sample_id)` are labeled as ACCESS-only.
* Variants at positions with coverage less than 100 were excluded
* We are plotting VAF+0.0001 for the log plots. 

## Copy number alterations
```{r cna, fig.height=3, fig.width=9}
if(nrow(cna)>0){
  fig2<-cna_plot(cna, xlimits, xbreaks)
  subplot(fig1,fig2,nrows=2,shareX=TRUE, heights=c(0.2,0.7), which_layout=1)
}
```


## Clonal SNVs/INDELs VAF adjusted for copy number {.tabset}

### Linear
```{r adjustedvaf-linear, fig.height=4}
clonal<-subset(final,final$clonality=="CLONAL")
if(nrow(clonal)>0){
  clonal$ncn <- 2
  clonal[which((clonal$Chromosome=="X" | clonal$Chromosome=="Y") & clonal$sex == "M"),"ncn"]<-1
  clonal$adjustedvaf<-clonal$vaf*clonal$ncn / (clonal$expected_alt_copies + (clonal$ncn - clonal$tcn)*clonal$vaf)
  clonal$adjustedvaf<-round(clonal$adjustedvaf,4)
  clonaltoplot<-clonal
  if(length(unique(clonal$VarName))>1){
   clonaltoplot<-rbind(clonal,data.frame(clonal[,c("vaf","adjustedvaf","collection_date", "cmo_sample_id_plasma")] %>% group_by(cmo_sample_id_plasma,collection_date) %>% summarize_all(list(mean)) %>% mutate(clonality="CLONAL", VarName="average")),fill=TRUE)
   clonaltoplot$vaf<-round(clonaltoplot$vaf,4)
   clonaltoplot$adjustedvaf<-round(clonaltoplot$adjustedvaf,4)
  }
  fig2<-vaf_plot(clonaltoplot, xlimits, xbreaks, varcolors, yaccuracy=0.01, log=FALSE, cnadjusted = TRUE)
  subplot(fig1,fig2,nrows=2,shareX=TRUE, heights=c(0.2,0.8))
}
```

### Log
```{r adjustedvaf-log, fig.height=4}
if(nrow(clonal)>0){
  fig2<-vaf_plot(clonaltoplot, xlimits, xbreaks, varcolors, yaccuracy=0.01, log=TRUE, cnadjusted = TRUE)
  subplot(fig1,fig2,nrows=2,shareX=TRUE, heights=c(0.2,0.8))
}
```

### Table
```{r snv_clonal_table}
print_table(clonal[,c("cmo_sample_id_plasma","collection_date","VarName","tcn","expected_alt_copies","ncn", "vaf","adjustedvaf")])
```

### Description

We adjust the variant allele fractions to account for the copy number alterations of the segments they are in. \
Since it is not easy to call copy number changes from ACCESS data, here we rely on the copy number alterations called by FACETS in the IMPACT sample `r toString(impact_sample_id)`.\
*Note: This assumes that there are no changes to copy numbers of these segments between the IMPACT and ACCESS samples.* \


$$VAF = \frac{T_{ALT}*P}{T_{CN}*P + N_{CN}*(1-P)}$$

where

* VAF = variant allele fraction (expected)
* P = tumor purity
* T~ALT~ = alternate copies in tumor
* T~CN~ = total copies in tumor
* N~CN~ = total copies in normal

By solving for this, we get:
$$P = \frac{N_{CN}*VAF}{T_{ALT} + (N_{CN} - T_{CN})*VAF}$$

For ACCESS samples, we compute this value and call it adjusted VAF (VAF~adj~).
